package.json

```diff
@@ -20,32 +21,33 @@
     "glob": "^7.1.3",
     "js-yaml": "^3.13.1",
     "json-pointer": "^0.6.2",
-    "json-refs": "^3.0.13",
+    "json-refs": "^3.0.15",
```

Comment:
Why are you bumping all these patch versions?  Do we need fixes only in the newer versions?  If not, you shoould leave these alone and only update `package-lock.json`.

Reply:
@mikeharder what is the downside of updating `package.json` ?

Reply:
For a reusable package (like `@azure/core-rest-pipeline`), it's bad since it artificially constrains the dependency versions you are compatible with.  The ecosystem works best when most packages float from the lowest version they are actually compatible with (e.g. `^2.0.0`).  To minimize cases where packages have incompatible dep ranges, and you need to install multiple versions side-by-side.

However, to guarantee correctness, this requires "min-max testing", where you run tests twice, once with the lowest and once with the highest compatible versions.  Otherwise you can accidentally take a dependency on an API added in a new minor version, and break consumers who pinned the older minor version.

So for a package that's more of a "standalone application" like `openapi-diff`, where it's not worth doing min-max testing, I think you're correct that either of these strategies would be better:

1. When publishing a new version of your package, update all deps in `package.json` to float starting from their current version (e.g. `^1.2.3`).  This guarantees you aren't using an API from a newer version, since there is no newer version.

2. Float with `~` instead of `^`, and only update when you need to use an API in a newer version (or if a dep has a security fix, etc).  If your dependencies follow semver correctly, new APIs should bump the minor version, which will require you to update your dep version before you can call the new API.

Between these two options, I would say 1 is more work to keep updated, but it should minimize the issue of side-by-side deps if someone does use you as a reusable package instead of an app.  The main advantage of option 2 is less churn in `package.json`.

`@typespec/compiler` appears to use option 2, though I'm not sure how intentionally.

I'm fine using option 1 for openapi-diff (as you did here).  But we'd also probably be fine to just float the lowest possible (e.g. `^1.0.0` instead of `^1.2.3`).  In practice, the majority of consumers will just install the latest floated versions of everything anyway, so it won't even matter.

Reply:
@mikeharder awesome, thank you for the detailed explanation! I added your comment to my personal reference of valuable GH comments :)

Reply:
In practice, none of this probably matters for `oav`.  If we want to update `package.json` periodically or every time we ship, fine.  If we want less churn, only update dep versions when needed.